### PR TITLE
feat: next 15

### LINKED ISSUE #1812: bug: When NextAuth authentication provider is selected, the Next Steps section need to indicate setting up Discord client info in .env file
### Provide environment information

  System:
    OS: macOS 11.7.10
    CPU: (8) x64 Intel(R) Core(TM) i7-4770HQ CPU @ 2.20GHz
    Memory: 148.06 MB / 16.00 GB
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 21.7.1 - /usr/local/bin/node
    Yarn: 1.22.17 - /usr/local/bin/yarn
    npm: 10.5.0 - /usr/local/bin/npm
    pnpm: 8.15.5 - ~/Library/pnpm/pnpm

"ct3aMetadata": {
    "initVersion": "7.30.0"
  }

### Describe the bug

npm run dev

> test-app@0.1.0 dev
> next dev

❌ Invalid environment variables: {
  DISCORD_CLIENT_ID: [ 'Required' ],
  DISCORD_CLIENT_SECRET: [ 'Required' ]
}

Node.js v21.7.1

### Reproduction repo

Part of setting up the project nothing to deploy yet.

### To reproduce

Ran `npm create t3-app@latest` to create new app with the following selections:

>│
> ◇  What will your project be called?
> │  test-app
> │
> ◇  Will you be using TypeScript or JavaScript?
> │  TypeScript
> │
> ◇  Will you be using Tailwind CSS for styling?
> │  Yes
> │
> ◇  Would you like to use tRPC?
> │  Yes
> │
> ◇  What authentication provider would you like to use?
> │  NextAuth.js
> │
> ◇  What database ORM would you like to use?
> │  Prisma
> │
> ◇   EXPERIMENTAL  Would you like to use Next.js App Router?
> │  Yes
> │
> ◇  What database provider would you like to use?
> │  SQLite
> │
> ◇  Should we initialize a Git repository and stage the changes?
> │  Yes
> │
> ◇  Should we run 'npm install' for you?
> │  Yes
> │
> ◇  What import alias would you like to use?
> │  ~/
> 
> Using: npm

Followed next steps provided:
  cd test-app
  npm run db:push
  npm run dev

### Additional information

Suggestion to enhance next steps section to include something about setting environment variables when using NextAuth.

### LINKED ISSUE #1979: feat: AuthJS v5
### Is your feature request related to a problem? Please describe.

What is still outstanding here? Are we just waiting for them to list the next JS flavor as not in beta?

### Describe the solution you'd like to see

AuthJS v5 implementation

### Describe alternate solutions

NextAuth v4What is still outstanding here? Are we just waiting for them to list the next JS flavor as not in beta

### Additional information

_No response_

### LINKED ISSUE #1999: feat: Will the Next.js version consider updating to 15?
### Is your feature request related to a problem? Please describe.

Next.js 15 is officially stable, and the official blog states that it can be used in production environments. Are there any plans to follow up with the official version?

### Describe the solution you'd like to see

Next.js updated to 15

### Describe alternate solutions

Refer to the update details in[Next.js 15 | Next.js --- Next.js 15](https://nextjs.org/blog/next-15)to check for breaking changes, modify the current code, and migrate to Next.js 15.

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by ronanru**
## ✅ Checklist

- [x] I have followed every step in the [contributing guide](https://github.com/t3-oss/create-t3-app/blob/main/CONTRIBUTING.md) (updated 2022-10-06).
- [x] The PR title follows the convention we established [conventional-commit](https://www.conventionalcommits.org/en/v1.0.0/)
- [x] I performed a functional test on my final commit

---

## Changelog

1) Update to Next 15
2) Update `next-auth` to v5 (beta)

Closes #1979
Closes #1999 
Closes #1812

### CODE REVIEW

**Comment by juliusmarminge** [comment]
How useful really is this considering the cjs limitations?

**Comment by ronanru** [comment]
> How useful really is this (ts config) considering the cjs limitations?

I don't like seeing js files in my src folder. With ts next config `env.js` can be `env.ts`. I don't think it changes anything else.

`next.config.ts` should be more useful in `create-t3-turbo`. I think we can remove `jiti` dependency there, when we update to next 15 there.

**Comment by juliusmarminge** [comment]
Are you sure this works? t3-env is an ESM only package which is called out to be incompatible image

**Comment by ronanru** [comment]
> Are you sure this works?
I tried multiple configuration, everything seems to work. Dev, Build, LSP all work. I get an error if for env variables are missing.

**Comment by juliusmarminge** [comment]
Try doing something esm in it, it will crash

This works: 

```ts
// app/env.js
import { createEnv } from "@t3-oss/env-nextjs";
import { vercel } from "@t3-oss/env-nextjs/presets"
import { z } from "zod";

console.log(import.meta.filename)

export const env = createEnv({
  extends: [vercel()],
  server: {
    FOO: z.string()
  },
  experimental__runtimeEnv: {
    FOO: "bar"
  }
});

// next.config.js
import "./app/env.js"

export default {
  /* config options here */
};
```

CleanShot 2024-10-22 at 17 06 27

Change file extensions to `.ts`. It crashes:

CleanShot 2024-10-22 at 17 06 51

**Comment by ronanru** [comment]
oh, ok. I reverted the next config changes. We can add them later when the ESM thing is fixed.

File: cli/src/installers/dependencyVersionMap.ts [hunk 1]
```diff
@@ -4,7 +4,7 @@
  */
 export const dependencyVersionMap = {
   // NextAuth.js
-  "next-auth": "^4.24.7",
+  "next-auth": "beta",
```
Comment (juliusmarminge) [inline cli/src/installers/dependencyVersionMap.ts:7]: should probably pin this while it's still in beta.

```suggestion
  "next-auth": "5.0.0-beta.25",
```

File: cli/template/extras/src/server/auth/base.ts [hunk 2]
```diff
@@ -33,36 +29,20 @@ declare module "next-auth" {
...
+export const { auth, handlers, signIn, signOut } = NextAuth({
+  providers: [
+    DiscordProvider({
+      clientId: env.DISCORD_CLIENT_ID,
+      clientSecret: env.DISCORD_CLIENT_SECRET,
+    }),
+  ],
   callbacks: {
-    session: ({ session, token }) => ({
+    session: ({ session, user }) => ({
```
Comment (juliusmarminge) [inline cli/template/extras/src/server/auth/base.ts:40]: user is not avaialble with jwt session type afaik, or have they changed this in recent versions?

File: cli/template/extras/src/server/auth/with-prisma.ts [hunk 3]
```diff
@@ -46,27 +48,4 @@ export const authOptions: NextAuthOptions = {
...
-    }),
-    /**
-     * ...add more providers here.
-     *
-     * Most other providers require a bit more work than the Discord provider. For example, the
-     * GitHub provider requires you to add the `refresh_token_expires_in` field to the Account
-     * model. Refer to the NextAuth.js docs for the provider you want to use. Example:
-     *
-     * @see https://next-auth.js.org/providers/github
-     */
```
Comment (juliusmarminge) [inline cli/template/extras/src/server/auth/with-prisma.ts:63]: bring back these comments. they are for educational purposes

**Review by juliusmarminge — CHANGES_REQUESTED** [review]
To make sure the auth setup works even if someone has a middleware file, let's use the pattern with separate config and intitialization, os e.g. `src/auth/config.ts` defines the core config and `src/auth/index.ts` inits next auth. also export the auth function wrapped in `React.cache` so it's deduped.

See here for example: https://github.com/juliusmarminge/trellix-trpc/tree/main/src/auth, or https://github.com/t3-oss/create-t3-turbo/tree/main/packages/auth/src

**Review by juliusmarminge — APPROVED** [review]
I made some more changes to get it more in line with Next Auth v5 standards.

I did some manual testing but would be nice with some more before merging. Can you look it over again @ronanru and make sure my additions are all good?

**Comment by TheCukitoDev** [comment]
Can @t3dotgg @c-ehrlich and @nexxeln please review the code please?? I think it would be useful to have it implemented already

**Comment by ronanru** [comment]
Just tested your changes, everything looks good @juliusmarminge

**Review by lacymorrow — APPROVED** [review]
LGTM++

**Pull request merged**